tests/evm/test_balance.py

```diff
@@ -0,0 +1,120 @@
+import pytest
+
+from zkevm_specs.evm import (
+    AccountFieldTag,
+    Block,
+    Bytecode,
+    CallContextFieldTag,
+    ExecutionState,
+    RWDictionary,
+    StepState,
+    Tables,
+    verify_steps,
+)
+from zkevm_specs.util import (
+    EXTRA_GAS_COST_ACCOUNT_COLD_ACCESS,
+    GAS_COST_WARM_ACCESS,
+    RLC,
+    U160,
+    U256,
+    rand_address,
+    rand_bytes,
+    rand_fq,
+    rand_range,
+    rand_word,
+)
+
+TESTING_DATA = [
+    (0x30000, 0, True, True, True),
+    (0x30000, 0, True, False, True),
+    (0x30000, 200, False, True, True),
+    (0x30000, 200, False, False, True),
+    (
+        rand_address(),
+        rand_word(),
+        rand_range(2) == 0,
+        rand_range(2) == 0,
+        True,  # persistent call
+    ),
+    (
+        rand_address(),
+        rand_word(),
+        rand_range(2) == 0,
+        rand_range(2) == 0,
+        False,  # reverted call
+    ),
+]
+
+
+@pytest.mark.parametrize("address, balance, is_non_existing, is_warm, is_persistent", TESTING_DATA)
+def test_balance(
+    address: U160, balance: U256, is_non_existing: bool, is_warm: bool, is_persistent: bool
+):
+    randomness = rand_fq()
+
+    result = 0 if is_non_existing else balance
+
+    tx_id = 1
+    call_id = 1
+
+    rw_counter_end_of_reversion = 0 if is_persistent else 9
+    reversible_write_counter = 0
+
+    rw_table = set(
+        RWDictionary(1)
+        .stack_read(call_id, 1023, RLC(address, randomness))
+        .call_context_read(tx_id, CallContextFieldTag.TxId, tx_id)
+        .call_context_read(
+            tx_id, CallContextFieldTag.RwCounterEndOfReversion, rw_counter_end_of_reversion
+        )
+        .call_context_read(tx_id, CallContextFieldTag.IsPersistent, is_persistent)
+        .tx_access_list_account_write(
+            tx_id,
+            address,
+            True,
+            is_warm,
+            rw_counter_of_reversion=rw_counter_end_of_reversion - reversible_write_counter,
```

Comment:
I tried to review [reversible-write-reversion doc](https://github.com/privacy-scaling-explorations/zkevm-docs/blob/main/src/design/reversible-write-reversion.md). It seems that `reversible_write_counter` should be initialized to `0`.

And I referenced the code of [test_extcodecopy.py](https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/tests/evm/test_extcodecopy.py#L77) and [test_extcodehash.py](https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/tests/evm/test_extcodehash.py#L69).

Reply:
I think in this test the `rw_counter_end_of_reversion` and `reversible_write_counter` are not super relevant because we're only checking the constraints of the `BALANCE` step with the `STOP/REVERT` step as next.  But we're not checking the constraints of the `STOP/REVERT` step.
The constraints related to `rw_counter_end_of_reversion` and `reversible_write_counter` are checked during the `REVERT` step, so even if we have wrong values here the test would pass.

As @silathdiir `reversible_write_counter` should start at 0 when a call begins.  I think `rw_counter_end_of_reversion` should be 9 + the count of reversible operations that happened (which I think should be 1 for the account access list write).

Reply:
Fixed `rw_counter_end_of_reversion` to `10` with comment:
```
# 9 + 1 reversible operation (the account access list write)
rw_counter_end_of_reversion = 0 if is_persistent else 10
```